File: python/day23part2.py
#!/usr/bin/python3
#
# My working for Day 23 Part 2
#
# PROBLEM: Play Crab Cups with some made-up complexities -
#          a million cups over ten million rounds.
#
#   Given an input list of INTs, these represent a "circle" of cups clockwise
#   from the first in the list (modulo arithmetic). The game rules are:
#      * First cup becomes CURRENT.
#      * Cups 2,3,4 are REMOVED from list.
#      * DESTINATION cup selected as <Current LABEL>-1
#            * If DESTINATION was in the REMOVED list, select <Current Label>-2
#              (repeat until matched. Wrap-around as required)
#      * <REMOVED> are then placed immediately *clockwise* of DESTINATION in
#        their original order
#      * CURRENT is selected as next in the list.
import time

cuplist="418976235"
#Demo data
#cuplist="389125467"

debug=False

doBigList=True

#A global dictionary. Order the cups by "which cup is next"
#(Confession: I needed hints to get this far. I have solutions using pop/insert)
#(but list / dict manipulation of 1million elements is too slow)
cupLinkedList={}

#Some global metadata. Expect "min", "max" and "num"
cMeta={}

################################################################################
#Play the game via the miracle of linked lists.
def play_crabcups_ll(current) :
    global cupLinkedList
    global cMeta
    #Because the LL is naturally circular, this is easy. As long as you remember
    #to move current's pointer.....
    removed=[]
    nr=cupLinkedList[current]
    for i in range(1,4) :
        removed.append(nr)
        nr=cupLinkedList[nr]
    cupLinkedList[current]=nr
    if debug: print(f" current={current}, removed={removed}")

    #Work out what the "result" cup is. All we need is the min, max and removed list.
    founddest=False
    dest_cup_label=current
    while not founddest :
        if debug : print(f"{dest_cup_label} > {cMeta['min']} ?")
        if dest_cup_label > cMeta['min'] :
            dest_cup_label=dest_cup_label - 1
        else :
            dest_cup_label=cMeta['max']
        if dest_cup_label not in removed :
            founddest=True

    if debug: print(f"Dest: {dest_cup_label}")

    #Now we need to re-arrange the indexes around this dest label
    org_dest_link = cupLinkedList[dest_cup_label]
    nlink=dest_cup_label
    for c in removed :
        cupLinkedList[nlink] = c
        nlink=c
    cupLinkedList[nlink]=org_dest_link

    #Our return value is now simply the value that comes after the one we
    #started with
    return cupLinkedList[current]

################################################################################
# See if you can guess why I need this....
def sanity_check_linked_list(checkmylist) :
    expectedHops=len(checkmylist.keys())
    if debug : print(f"VALIDATING Linked List {checkmylist}")
    posZero = checkmylist[1] #Dodgy assumption but we'll make it anyway
    hops=1
    posFinal = checkmylist[posZero]
    if debug : print(f"\tFirst Hop goes {posZero}->{posFinal}")
    while hops<(expectedHops-1):
        if debug : print(f"\tHop {hops} goes from {posFinal}",end='->')
        posFinal=checkmylist[posFinal]
        if debug : print(posFinal)
        if posFinal == posZero :
            print(f"ERROR Validating linked list. Reached {posZero} after only {hops} steps instead of {expectedHops}")
            return False
        hops+=1
    if checkmylist[posFinal] == posZero :
        print(f"Linked List validated OK, {hops} steps got us back to {posZero}")
        return True
    else :
        print(f"Linked list failed validation; {hops} steps didn't get back to {posZero}")
        return False

################################################################################
def partOneAnswer(llist,start) :
    n=llist[start]
    out=""
    while n != start :
        out+=str(n)
        n=llist[n]
    return str(n)+out

################################################################################
################################################################################
################################################################################
if __name__ == '__main__':
    #max/min values useful to cache to enable wrap-around
    cuplist=[int(i) for i in list(cuplist)]

    cMeta['min']=min(cuplist)
    cMeta['max']=max(cuplist)
    #The internet tells me the right way to do this is a dict of cupbylabel->nextcup
    cupZero = cuplist[0]
    cupLinkedList[cupZero] = cuplist[1]
    for c in range(2,len(cuplist)) :
        cupLinkedList[cuplist[c-1]] = cuplist[c]

    v=cuplist[len(cuplist)-1]

    if doBigList :
        if debug: print(f"Extending {cupLinkedList} to 1,000,000 entries....")
        vNext=cMeta['max']+1
        for c in range(len(cupLinkedList.keys())-1,999998) :
            cupLinkedList[v]=vNext
            cMeta['max']=vNext
            v=vNext
            vNext+=1

    #Loop back to zero
    cupLinkedList[v]=cupZero
    cMeta['num']=len(cupLinkedList.keys())

    if debug :
        print(f"Last V = {v}; links back to {cupLinkedList[v]}")
        for c in list(cupLinkedList.keys())[0:20] :
            print(f"{c} -> {cupLinkedList[c]}")
        if debug: print(f"Cuplist metadata: {cMeta}")

    round=0
    maxRounds=10000000

    current=cuplist[0]
    tStart = time.time()
    while round < maxRounds :
        round+=1
        if debug:
            print("-"*20, "ROUND ", round, " ", "-"*20)
            print(f"{round} : ",end='')
            print(partOneAnswer(cupLinkedList,current),end="->")
        current = play_crabcups_ll(current)
        if debug: print(partOneAnswer(cupLinkedList,current))

    tEnd = time.time()

    tDiff = tEnd - tStart
    tRate = maxRounds / tDiff
    ttComplete = tDiff * (10000000 / maxRounds)
    print(f"(elapsed time: {tDiff:3.3f}, turns/sec: {tRate:1.2f})")
    print(f"Estimated time to complete 100,000,000 rounds: {ttComplete:1.2f} seconds")
    print(f"({ttComplete/3600:1.2f} Hours; {ttComplete/86400:1.2f} Days)")

    #Sanity check the resulting list
    if sanity_check_linked_list(cupLinkedList) :
        #Thanks to this being linked list, working out what's after 1 is stupidly easy:
        ansp1 = cupLinkedList[1]
        ansp2 = cupLinkedList[ansp1]
        ans_product = ansp1 * ansp2

        print(f"The numbers next to cup 1 after {round} rounds are {ansp1} and {ansp2}")
        print(f"The answer to Part Two is therefore {ans_product}")

        #Part one answer:
        #print(f" Part One list: {cupLinkedList}")
        #p1answer=partOneAnswer(cupLinkedList,1)
        #print(f" Part One Answer: {p1answer}")

File: go/day23part2.go
package main
/*
# PROBLEM: Play Crab Cups with some made-up complexities -
#          a million cups over ten million rounds.
#
#   Given an input list of INTs, these represent a "circle" of cups clockwise
#   from the first in the list (modulo arithmetic). The game rules are:
#      * First cup becomes CURRENT.
#      * Cups 2,3,4 are REMOVED from list.
#      * DESTINATION cup selected as <Current LABEL>-1
#            * If DESTINATION was in the REMOVED list, select <Current Label>-2
#              (repeat until matched. Wrap-around as required)
#      * <REMOVED> are then placed immediately *clockwise* of DESTINATION in
#        their original order
#      * CURRENT is selected as next in the list.

(Adapted from my Python answer, as an attempt to do large maps in Go)

*/
import (
  "fmt"
  //"log"
  "time"
)

//------------------------------------------------------------------------------
func build_cup_linkedlist(initCups []int, totalLength int)  map[int]int {
    cuplist := make(map[int]int)
    //Scheme is: Build the initial list from the string of numbers passed in,
    //expand to the length requested linking n-1->n except for the last which
    //links to the first.
    cupZero:=initCups[0]
    cMin:=99999999
    cMax:=0
    prevC:=0
    for i,dv := range initCups {
      //Bookkeeping min/max
      if(dv<cMin) { cMin=dv}
      if(dv>cMax) { cMax=dv}
      if i==0 {
        cupZero = dv
        prevC=dv
      //General case - cup n-1 links to cup n
      } else {
        cuplist[prevC]=dv
        prevC=dv
      }
    }
    defLen:=len(cuplist)+1
    //And will run to make up <totalLength> entries:
    for i := defLen+1; i<=totalLength; i++ {
      cuplist[prevC]=i
      prevC=i
    }
    //link final element back to cupZero
    cuplist[len(cuplist)+1] = cupZero

    return cuplist
}
//------------------------------------------------------------------------------
func sanity_check_linked_list(ll map[int]int, fromStart int) bool {
  //List is valid if LAST element links back to FIRST element, and following
  //the chain goes through EVERY element in the list exactly once.
  hopCount:=1
  expectedHops:=len(ll)
  cHop := ll[fromStart]
  for cHop != fromStart && hopCount<=expectedHops {
    nHop := ll[cHop]
    //fmt.Printf("hop %d:  %d -> %d\n", hopCount,cHop,nHop)
    hopCount++
    if nHop == fromStart && hopCount!=expectedHops {
      fmt.Printf("Loop linked to start (%d) after only %d of %d expected hops\n",
        fromStart,
        hopCount,
        expectedHops)
      return false //Linked to start too soon
    }
    cHop = nHop
  }

  if cHop == fromStart && expectedHops==hopCount {
      return true
  }
  fmt.Printf("Loop didn't link back to start (%d) after expected %d hops, only %d taken\n",
      fromStart,
      expectedHops,
      hopCount)
  return false
}

//------------------------------------------------------------------------------
// MAPS are always pass-by-reference so there's no funky pointer arithmetic
//      required. We always end up modifying the global value. A benefit, in this
//      case...
func crab_cups_round(cuplist map[int]int, currentCup,cMin,cMax int) int {
  var removed [3]int
  removed[0]=cuplist[currentCup]
  removed[1]=cuplist[removed[0]]
  removed[2]=cuplist[removed[1]]
  ptrNext := cuplist[removed[2]]
  //point the current entry "beyond" the removed cups:
  cuplist[currentCup] = ptrNext

  //Apply the crab-cups rule: "target" is current-1 with wrap-around if needed
  foundTarget:=false
  destCup:=currentCup
  for !foundTarget {
    if destCup>cMin {
      destCup = destCup-1
    } else {
      destCup = cMax
    }
    //Check this value hasn't been removed
    if removed[0]!=destCup && removed[1]!=destCup && removed[2]!=destCup {
      foundTarget=true
    }
  }

  //Update the linkage by re-inserting the removed AFTER target cup.
  orgPtr := cuplist[destCup]
  cuplist[destCup] = removed[0]
  cuplist[removed[0]] = removed[1]
  cuplist[removed[1]] = removed[2]
  cuplist[removed[2]] = orgPtr

  //And finish by returning whatever is now after the current cup
  return cuplist[currentCup]

}

//------------------------------------------------------------------------------
//------------------------------------------------------------------------------
//------------------------------------------------------------------------------
func main() {
  //Live data:
  initCups := []int{4,1,8,9,7,6,2,3,5}
  //(Actual answer: 563362809504)
  //Test data:
  //(Expected result: 1 next to 934001,159792. Answer: 149245887792)
  //initCups := []int{3,8,9,1,2,5,4,6,7}

  preInit := time.Now()
//  cuplist := build_cup_linkedlist(initCups,1000000)
  cuplist := build_cup_linkedlist(initCups,1000000)
  postInit := time.Now()
  initElapsed := postInit.Sub(preInit)
  fmt.Printf("Initialising the Cups List to %7d entries took %3.2f seconds\n",
             len(cuplist),
             (initElapsed.Seconds() * 1e6) /1e6)
  preCheck := time.Now()
  isValidChain := sanity_check_linked_list(cuplist,initCups[0])
  postCheck := time.Now()
  chkElapsed := postCheck.Sub(preCheck)
  fmt.Printf("Sanity-checking initial list took %3.2f seconds and isvalid=%t\n",
    (chkElapsed.Seconds() * 1e6 ) / 1e6,
    isValidChain)
  //re-calculate the min and max values in the cuplist.
  cupMin:=9999999999999
  cupMax:=0
  for _,v := range cuplist {
    if v>cupMax { cupMax=v}
    if v<cupMin { cupMin=v}
  }

  //Now play the game the proscribed number of turns
  maxRounds := 10000000
  current:=initCups[0]
  tStart := time.Now()
  for i := 0; i<maxRounds; i++ {
    current = crab_cups_round(cuplist,current,cupMin,cupMax)
  }
  tEnd := time.Now()
  tDiff := tEnd.Sub(tStart)
  tRate := int64(maxRounds*1000000) / tDiff.Microseconds()
  validpostExecution:=sanity_check_linked_list(cuplist,initCups[0])
  fmt.Printf("%d rounds of Crab Cups took %3.2f or %d rounds/sec and the result is valid: %t\n",
    maxRounds,
    (tDiff.Seconds() * 1e6)/1e6,
    tRate,
    validpostExecution)

  iNext:=cuplist[1]
  iNextPlus:=cuplist[iNext]
  fmt.Printf("The numbers after 1 are %d and %d\n",iNext,iNextPlus)

  partTwoAnswer := iNext * iNextPlus
  fmt.Printf("The Part Two answer is therefore: %d\n",partTwoAnswer)


}
